rbd: add EnsureImageCleanup to ensure image cleanup from trash

Rakshith-R opened the pull request:
After moving moving image to trash, if `trash remove` step fails,
then external-provisioner will issue subsequent requests, in which
image will be absent in pool( will be in trash) and omap cleanup will
be done with stale image left in trash with no `trash remove` step on it.

To avoid this scenario list trash images and find corresponding id for given
image name and add a task to flatten when we encounter a ErrImageNotFound.

Fixes: ceph#1728

Rakshith-R committed:
rbd: add EnsureImageCleanup to ensure image cleanup from trash

After moving moving image to trash, if `trash remove` step fails,
then external-provisioner will issue subsequent requests, in which
image will be absent in pool( will be in trash) and omap cleanup will
be done with stale image left in trash with no `trash remove` step on it.

To avoid this scenario list trash images and find corresponding id for given
image name and add a task to flatten when we encounter a ErrImageNotFound.

Fixes: #1728

Signed-off-by: Rakshith R <rar@redhat.com>

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -564,6 +564,16 @@ func deleteImage(ctx context.Context, pOpts *rbdVolume, cr *util.Credentials) er
 
 		return err
 	}
+	defer func() {
```
Indention looks really weird in GitHub while reviewing this change... I wonder if there is some strange space/tab mismatch here.

Rakshith-R replied:
>Indention looks really weird in GitHub while reviewing this change... I wonder if there is some strange space/tab mismatch here.

Double checked, looks fine to me. Maybe because of nesting depth, it looks that way.

nixpanic reviewed (commented):
Looks good in general, but what is the expectation towards the user when the image is restored from trash? As the encryption passphrase has been deleted, the image can not be used anymore... The expectations should probably be documented.

Rakshith-R commented:
> Looks good in general, but what is the expectation towards the user when the image is restored from trash? As the encryption passphrase has been deleted, the image can not be used anymore... The expectations should probably be documented.

@nixpanic 
There's actually no way to recover/cancel a deletion request on a PVC as far as my findings go.
This image is certianly to be deleted in subsequent request from external-provisioner(which will definitely be fired).

nixpanic commented:
On Mon, Aug 23, 2021 at 05:53:13AM -0700, Rakshith R wrote:
> > Looks good in general, but what is the expectation towards the user when the image is restored from trash? As the encryption passphrase has been deleted, the image can not be used anymore... The expectations should probably be documented.
> 
> @nixpanic 
> There's actually no way to recover/cancel a deletion request on a PVC as far as my findings go.
> This image is certianly to be deleted in subsequent request from external-provisioner(which will definitely be fired).

Right, that is what I assumed too. Documenting this in the commit
message or in the comments of the function would be good.

Rakshith-R commented:
> Right, that is what I assumed too. Documenting this in the commit message or in the comments of the function would be good.

Documented the behavior in both the comment and commit message. Hope it is clearer now.

Madhu-1 commented on internal/rbd/rbd_util.go:
```diff
@@ -564,6 +564,19 @@ func deleteImage(ctx context.Context, pOpts *rbdVolume, cr *util.Credentials) er
+		// If `trash remove` step fails, then in subsequent request, image should be
+		// found in the pool for deletion to resume normally. Hence, Restore the image
+		// from trash on err from `trash remove` step.
+		if err != nil {
```
Thanks for giving it a try, but this fix will still leave the stale images if the function inside defer fails or never gets called due to restart, etc

The way I was thinking to fix this one was

* If the image Not found in the pool do below operations as part of delete
* As we store the imageID check do we have any image in the trash with this ID
* In some cases we don't store ID and we just move the image to the trash in that case check do we have an image with a name in the Trash
* If the Image is found add a task to remove it from the trash
* If the image is not found return notFound or success as expected.

Rakshith-R replied:
> Thanks for giving it a try, but this fix will still leave the stale images if the function inside defer fails or never gets called due to restart, etc

This mainly did focus on just errors from `trash remove` api calls.
 pod restart at a particular step is highly unlikely.

> 
> The way I was thinking to fix this one was
> 
> * If the image Not found in the pool do below operations as part of delete
> * As we store the imageID check do we have any image in the trash with this ID

We can directly restore image from trash here with imageID(if imageID != "") 
if found, send back an error and let another subsequent request handle it gracefully.
else, continue with cleanup.

> * In some cases we don't store ID and we just move the image to the trash in that case check do we have an image with a name in the Trash

IMO, we have been using and storing image ID for a very long time now and this is a cleanup bug patch and covering latest use cases is sufficient. 

The following steps increase complexity and add too many api calls.
(trash ls, find image, construct trash path, add task to remove  )

> * If the Image is found add a task to remove it from the trash
> * If the image is not found return notFound or success as expected.

@Madhu-1 , let me know what you think.

Madhu-1 replied:
> > Thanks for giving it a try, but this fix will still leave the stale images if the function inside defer fails or never gets called due to restart, etc
> 
> This mainly did focus on just errors from `trash remove` api calls.
> pod restart at a particular step is highly unlikely.


Not sure why we want to fix it half not the full as we know the bug. IMO we should not work on only one scenario

> > The way I was thinking to fix this one was
> > 
> > * If the image Not found in the pool do below operations as part of delete
> > * As we store the imageID check do we have any image in the trash with this ID
> 
> We can directly restore image from trash here with imageID(if imageID != "")
> if found, send back an error and let another subsequent request handle it gracefully.
> else, continue with cleanup.

Why do you want to restore the image which is already in the trash (Have you verified how much time restore can take
depends on the image size? if it takes less time we can consider that but that is not enough to handle all the corner cases)


> > * In some cases we don't store ID and we just move the image to the trash in that case check do we have an image with a name in the Trash
> 
> IMO, we have been using and storing image ID for a very long time now and this is a cleanup bug patch and covering latest use cases is sufficient.
> 

Are we storing the imageID of the temp clone which we create during the PVC-PVC clone? and also consider the case where image creation got succeeded but next operation like storing the imageID in omap fails etc

> The following steps increase complexity and add too many api calls.
> (trash ls, find image, construct trash path, add task to remove )
> 

Am open to better suggestion to fix it completely not partially.
 
> > * If the Image is found add a task to remove it from the trash
> > * If the image is not found return notFound or success as expected.
> 
> @Madhu-1 , let me know what you think.

Madhu-1 reviewed (commented):
Please add an E2E for this one by moving the image to the trash and try to delete the Volume.

Rakshith-R committed:
rbd: add EnsureImageCleanup to snapshot deletion

Signed-off-by: Rakshith R <rar@redhat.com>

Rakshith-R committed:
e2e: add test cases for EnsureImageCleanup

This tests pvc,pvcsmartclone,snapshot deletion when
underlying images are in trash.

Signed-off-by: Rakshith R <rar@redhat.com>

Madhu-1 commented on internal/rbd/rbd_util.go:
```diff
@@ -597,6 +597,28 @@ func deleteImage(ctx context.Context, pOpts *rbdVolume, cr *util.Credentials) er
+	return nil
+}
+
+func TrashRemoveImage(ctx context.Context, pOpts *rbdVolume, cr *util.Credentials) error {
```
don't export these functions keep it internal.

Madhu-1 commented on internal/rbd/rbd_util.go:
```diff
@@ -597,6 +597,28 @@ func deleteImage(ctx context.Context, pOpts *rbdVolume, cr *util.Credentials) er
+	return TrashRemoveImage(ctx, pOpts, cr)
+}
+
+func EnsureImageCleanup(ctx context.Context, pOpts *rbdVolume, cr *util.Credentials) error {
```
make this a method of rbdVolume?

Madhu-1 commented on e2e/rbd.go:
```diff
@@ -2761,6 +2761,110 @@ var _ = Describe("RBD", func() {
+				if err != nil {
+					e2elog.Failf("failed to delete pvc: %v", err)
+				}
+
```
add validation for `0` images in the pool

Rakshith-R commented:
@mergifyio rebase

Rakshith-R commented:
> @Mergifyio rebase


3 ci e2e failed due to timeout, #2264, http error and rebased on top of devel.
https://jenkins-ceph-csi.apps.ocp.ci.centos.org/job/mini-e2e-helm_k8s-1.20/2767/display/redirect
https://jenkins-ceph-csi.apps.ocp.ci.centos.org/job/mini-e2e-helm_k8s-1.21/1432/display/redirect
https://jenkins-ceph-csi.apps.ocp.ci.centos.org/job/mini-e2e_k8s-1.21/1447/display/redirect

@ceph/ceph-csi-contributors PTAL.

Rakshith-R commented:
/retest ci/centos/upgrade-tests-rbd

Rakshith-R commented:
> /retest ci/centos/upgrade-tests-rbd

failed due to #2264 

https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/upgrade-tests-rbd/detail/upgrade-tests-rbd/3366/pipeline

Rakshith-R commented:
/retest ci/centos/mini-e2e-helm/k8s-1.20

Rakshith-R commented:
> /retest ci/centos/mini-e2e-helm/k8s-1.20


panic: test timed out after 1h30m0s
https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e-helm_k8s-1.20/detail/mini-e2e-helm_k8s-1.20/2786/pipeline

Rakshith-R commented:
/retest ci/centos/mini-e2e/k8s-1.21

panic: test timed out after 1h30m0s
https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e_k8s-1.21/detail/mini-e2e_k8s-1.21/1466/pipeline

Madhu-1 commented on e2e/rbd.go:
```diff
@@ -2761,6 +2761,112 @@ var _ = Describe("RBD", func() {
+				defer func() {
+					err = deleteRBDSnapshotClass()
+					if err != nil {
+						e2elog.Failf("failed to delete VolumeSnapshotClass: %v", err)
```
`with error` is not required?

Madhu-1 commented on e2e/rbd.go:
```diff
@@ -2761,6 +2761,112 @@ var _ = Describe("RBD", func() {
+					_, stdErr, err = execCommandInToolBoxPod(f,
+						fmt.Sprintf("rbd snap purge %s %s", rbdOptions(defaultRBDPool), imageName), rookNamespace)
+					if err != nil || stdErr != "" {
+						e2elog.Logf(
```
make it as Failf or else the below command fails, it becomes hard to find the issue.

Madhu-1 commented on e2e/rbd.go:
```diff
@@ -2761,6 +2761,112 @@ var _ = Describe("RBD", func() {
+					if err != nil || stdErr != "" {
+						e2elog.Logf(
+							"failed to snap purge %s/%s: %v stdErr: %v",
+							defaultRBDPool,
```
log  `rbdOptions(defaultRBDPool)` instead of defaultRBDPool

Madhu-1 commented on e2e/rbd.go:
```diff
@@ -2761,6 +2761,112 @@ var _ = Describe("RBD", func() {
+						fmt.Sprintf("rbd trash move %s %s", rbdOptions(defaultRBDPool), imageName), rookNamespace)
+					if err != nil || stdErr != "" {
+						e2elog.Failf(
+							"failed to move rbd image %s/%s to trash: %v stdErr: %v",
```
log  `rbdOptions(defaultRBDPool)` instead of defaultRBDPool

Madhu-1 commented on e2e/rbd.go:
```diff
@@ -2761,6 +2761,112 @@ var _ = Describe("RBD", func() {
+
+				err = waitToRemoveImagesFromTrash(f, defaultRBDPool, deployTimeout)
+				if err != nil {
+					e2elog.Failf("failed to validate rbd images in pool %s trash with error %v", defaultRBDPool, err)
```
same here as  well log  `rbdOptions(defaultRBDPool)` instead of defaultRBDPool? and remove `with error`?

Madhu-1 commented on internal/rbd/rbd_util.go:
```diff
@@ -565,6 +565,24 @@ func (rv *rbdVolume) getTrashPath() string {
+func (rv *rbdVolume) ensureImageCleanup(ctx context.Context) error {
+	trashInfoList, err := librbd.GetTrashList(rv.ioctx)
+	if err != nil {
+		log.ErrorLog(ctx, "failed to list images in trash: %w", err)
```
`%w` is not needed for logging it should be `%v`

Madhu-1 commented on internal/rbd/rbd_util.go:
```diff
@@ -597,6 +615,10 @@ func deleteImage(ctx context.Context, pOpts *rbdVolume, cr *util.Credentials) er
+	return trashRemoveImage(ctx, pOpts, cr)
+}
+
+func trashRemoveImage(ctx context.Context, pOpts *rbdVolume, cr *util.Credentials) error {
```
can we make this a method of rbdVolume? if we do that we can avoid passing the two parameters to the function

Rakshith-R replied:
> can we make this a method of rbdVolume? if we do that we can avoid passing the two parameters to the function

I doubt we copied cr into pOpts everywhere. 

I'll send a pr to use go-ceph rbd task api for flatten and trash remove, which will remove the need for passing `cr` everywhere and make it a method on rbdVolume then(+ some cleanup too).

Madhu-1 reviewed (changes requested):
few nits. changes LGTM

Madhu-1 commented on e2e/rbd.go:
```diff
@@ -2761,6 +2761,109 @@ var _ = Describe("RBD", func() {
+
+				pvcSmartClone, err := loadPVC(pvcSmartClonePath)
+				if err != nil {
+					e2elog.Failf("failed to load pvcSmartClone : %v", err)
```
`pvcSmartClone : %v` to `pvcSmartClone: %v`

Madhu-1 commented on e2e/rbd.go:
```diff
@@ -2761,6 +2761,109 @@ var _ = Describe("RBD", func() {
+
+				err = waitToRemoveImagesFromTrash(f, defaultRBDPool, deployTimeout)
+				if err != nil {
+					e2elog.Failf("failed to validate rbd images in trash %q: %v", rbdOptions(defaultRBDPool), err)
```
be consistent with formatting  use `%s` here

Rakshith-R commented:
@mergifyio rebase

humblec commented on internal/rbd/rbd_util.go:
```diff
@@ -565,6 +565,24 @@ func (rv *rbdVolume) getTrashPath() string {
 	return trashPath + "/" + rv.ImageID
 }
 
+func (rv *rbdVolume) ensureImageCleanup(ctx context.Context) error {
```
good to add some source code comment here

humblec commented on internal/rbd/rbd_util.go:
```diff
@@ -565,6 +565,26 @@ func (rv *rbdVolume) getTrashPath() string {
+// ensureImageCleanup finds image in trash and if found removes it
+// from trash.
+func (rv *rbdVolume) ensureImageCleanup(ctx context.Context) error {
+	trashInfoList, err := librbd.GetTrashList(rv.ioctx)
```
as we are calling `ensureImageCleanup` from more than caller, I think reading the trash list can be racy. Isnt it ?

Rakshith-R replied:
> as we are calling `ensureImageCleanup` from more than caller, I think reading the trash list can be racy. Isnt it ? 

Does not matter,  we are just checking if that particular image is in trash and if it is, add a task/ remove from trash. 

And delete operation on pvc/ snapshot is already protected by locks.

humblec replied:
so if its there and before the caller issue the task to remove from trash, if another thread did add a task , the first one going to fail?

Rakshith-R replied:
> so if its there and before the caller issue the task to remove from trash, if another thread did add a task , the first one going to fail? 

Due to delete op locking, two threads will never be acting on same image. 

If there is a task already added, new cmd to add task will be mapped to the same task without any errors or failures.

humblec replied:
yeah, even snap operation is on the lock, so should be safe here.

Rakshith-R commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

Rakshith-R commented:
> /retest ci/centos/mini-e2e-helm/k8s-1.22

failed due to #2264 
https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e-helm_k8s-1.22/detail/mini-e2e-helm_k8s-1.22/798/pipeline

Rakshith-R commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

Rakshith-R commented:
> /retest ci/centos/mini-e2e-helm/k8s-1.22

Failed due to #2264 
https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e-helm_k8s-1.22/detail/mini-e2e-helm_k8s-1.22/799/pipeline

nixpanic commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

nixpanic commented:
> /retest ci/centos/mini-e2e-helm/k8s-1.22

it seems there was some issue with the MON(s):
```
failed to validate rbd images in trash --pool=replicapool: failed to list images in trash 2021-10-20T13:52:59.494+0000 7fc1070f3700 -1 monclient: get_auth_request but no auth handler is set up
```
possibly the [logs](https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e-helm_k8s-1.22/detail/mini-e2e-helm_k8s-1.22/803/pipeline/110) contain more details

Rakshith-R commented:
@mergifyio rebase

Rakshith-R commented:
/retest ci/centos/mini-e2e-helm/k8s-1.20

Rakshith-R commented:
/retest ci/centos/mini-e2e/k8s-1.20

Rakshith-R commented:
/retest ci/centos/mini-e2e/k8s-1.20
